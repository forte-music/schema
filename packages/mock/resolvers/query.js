// @flow
import type { Connection, Playlist, PlaylistItem } from '../models';
import type { ConnectionArgs } from './types';

import { albums, artists, playlists, songs } from '../models';
import { mustGet } from '../utils';

const itemResolver = <T>(map: Map<string, T>) => (
  _: void,
  { id }: { id: string }
): T => mustGet(map, id);

const connectionResolver = <T>(map: Map<string, T>) => (
  _: void,
  args: ConnectionArgs
): Connection<T> =>
  handleConnection(
    Array.from(map.keys()).sort(),
    (key: string): T => mustGet(map, key),
    args
  );

const handleConnection = <InputType, NodeType>(
  keys: InputType[],
  getNode: (key: InputType) => NodeType,
  { first = 25, after }: ConnectionArgs
): Connection<NodeType> => {
  const lowerBound = after ? parseInt(after, 10) + 1 : 0;
  const upperBound = first === -1 ? keys.length : lowerBound + first;

  const acceptedKeys = keys.slice(lowerBound, upperBound);

  return {
    pageInfo: {
      count: keys.length,
      hasNextPage: upperBound <= keys.length,
    },
    edges: acceptedKeys.map((key, index) => ({
      cursor: (index + lowerBound).toString(),
      node: getNode(key),
    })),
  };
};

const queryResolvers = {
  Query: {
    album: itemResolver(albums),
    albums: connectionResolver(albums),

    artist: itemResolver(artists),
    artists: connectionResolver(artists),

    song: itemResolver(songs),
    songs: connectionResolver(songs),

    playlist: itemResolver(playlists),
    playlists: connectionResolver(playlists),
  },

  Playlist: {
    items: (
      { items }: Playlist,
      args: ConnectionArgs
    ): Connection<PlaylistItem> => handleConnection(items, item => item, args),
  },
};

export default queryResolvers;
