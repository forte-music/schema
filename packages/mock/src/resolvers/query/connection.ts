import { Connection } from '../../models';

export interface ConnectionArgs {
  after?: string;
  first?: number;
}

export const handleConnection = <InputType, NodeType>(
  keys: InputType[],
  getNode: (key: InputType) => NodeType,
  { first = 25, after }: ConnectionArgs
): Connection<NodeType> => {
  const lowerBound = after ? parseInt(after, 10) + 1 : 0;
  const upperBound = first === -1 ? keys.length : lowerBound + first;

  const acceptedKeys = keys.slice(lowerBound, upperBound);

  return {
    count: keys.length,
    pageInfo: {
      hasNextPage: upperBound < keys.length,
    },
    edges: acceptedKeys.map((key, index) => ({
      cursor: (index + lowerBound).toString(),
      node: getNode(key),
    })),
  };
};
