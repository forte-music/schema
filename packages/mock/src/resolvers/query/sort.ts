import { Connection } from '../../models';
import { ConnectionArgs, handleConnection } from './connection';

type SortBy = 'RECENTLY_ADDED' | 'LEXICOGRAPHICALLY' | 'RECENTLY_PLAYED';

interface SortParams {
  sortBy?: SortBy;
  reverse?: boolean;
  filter?: string;
}

interface Sortable {
  name: string;
  lastPlayed?: number;
  timeAdded: number;
}

interface LinkedSortInfo<T> {
  sortable: Sortable;
  original: T;
}

interface Comparator<T> {
  (a: T, b: T): number;
}

export const compare = <T>(a: T, b: T): number => {
  if (a > b) {
    return 1;
  }

  if (a === b) {
    return 0;
  }

  if (b > a) {
    return -1;
  }

  throw new TypeError(`Invariant a: ${a}, b: ${b}`);
};

export const reverseCompare = <T>(a: T, b: T): number => reverse(a, b, compare);

const reverse = <T>(a: T, b: T, inner: Comparator<T>): number => inner(b, a);

const identity = <T>(a: T, b: T, inner: Comparator<T>): number => inner(a, b);

// Returns how closely the record matches the query string.
const searchDistance = (record: string, query: string): number => {
  const normalizedQuery = query.toLowerCase().replace(/ /g, '');
  const normalizedRecord = record.toLowerCase();

  let strength = 0;
  let currentStrength = 0;
  let lastMatchIndex = 0;

  const khars = normalizedQuery.split('');
  for (const khar of khars) {
    const indexOfQueryCharacter = normalizedRecord.indexOf(
      khar,
      lastMatchIndex
    );
    if (indexOfQueryCharacter === -1) {
      // A character from the query isn't in the record. Sort to bottom.
      return -1;
    }

    if (lastMatchIndex === indexOfQueryCharacter - 1) {
      // continuation of last match
      currentStrength++;
    } else {
      // broken streak, reset streak strength
      currentStrength = 0;
    }

    // move last match index
    lastMatchIndex = indexOfQueryCharacter;

    // update strength
    strength += currentStrength;
  }

  return strength;
};

const sortWithReverse = <T>(
  reversed: boolean,
  inner: Comparator<T>
): Comparator<T> => {
  return (a: T, b: T): number => {
    const transform = reversed ? reverse : identity;
    return transform(a, b, inner);
  };
};

export type SortConnectionArgs = ConnectionArgs & {
  sort?: SortParams;
};

export const listItemsResolver = <T>(
  values: T[],
  { sort, ...connectionArgs }: SortConnectionArgs,
  process: (item: T) => Sortable
): Connection<T> => {
  const localHandleConnection = (items: T[]) =>
    handleConnection(items, value => value, connectionArgs);

  if (!sort) {
    // No sort was specified, use default sort.

    return localHandleConnection(values);
  }

  const { sortBy = 'LEXICOGRAPHICALLY', reverse = false, filter = '' } = sort;

  let sorter: Comparator<Sortable> = () => 0;
  let filterFunc = (item: Sortable) => searchDistance(item.name, filter) !== -1;

  switch (sortBy) {
    case 'RECENTLY_ADDED':
      sorter = (a: Sortable, b: Sortable) =>
        reverseCompare(a.timeAdded, b.timeAdded);
      break;

    case 'LEXICOGRAPHICALLY':
      sorter = (a: Sortable, b: Sortable) => compare(a.name, b.name);
      break;

    case 'RECENTLY_PLAYED':
      sorter = (a: Sortable, b: Sortable) =>
        reverseCompare(a.lastPlayed || 0, b.lastPlayed || 0);
      break;

    default:
      throw new TypeError(`Unknown Sort ${sortBy}`);
  }

  const mappedValues: LinkedSortInfo<T>[] = values.map(item => ({
    original: item,
    sortable: process(item),
  }));

  const filteredValues = mappedValues.filter(({ sortable }) =>
    filterFunc(sortable)
  );

  const sortedValues = filteredValues.slice();
  sortedValues.sort(
    sortWithReverse(
      reverse,
      (
        { sortable: a }: LinkedSortInfo<T>,
        { sortable: b }: LinkedSortInfo<T>
      ) => sorter(a, b)
    )
  );

  return localHandleConnection(sortedValues.map(({ original }) => original));
};

export const itemsResolver = <T extends Sortable>(map: Map<string, T>) => (
  _: void,
  args: SortConnectionArgs
): Connection<T> => {
  const values = Array.from(map.values());
  return listItemsResolver(values, args, item => item);
};
