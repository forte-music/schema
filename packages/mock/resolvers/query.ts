import { albums, artists, songs } from '../models/index';
import { mustGet } from '../utils';
import { itemsResolver } from './sort';

const itemResolver = <T>(map: Map<string, T>) => (
  _: void,
  { id }: { id: string }
): T => mustGet(map, id);

const queryResolvers = {
  Query: {
    album: itemResolver(albums),
    albums: itemsResolver(albums),

    artist: itemResolver(artists),
    artists: itemsResolver(artists),

    song: itemResolver(songs),
    songs: itemsResolver(songs),
  },
};

export default queryResolvers;
