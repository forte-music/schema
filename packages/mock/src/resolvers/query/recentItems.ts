import { Album, albums, Artist, artists } from '../../models/index';
import { reverseCompare } from './sort';

interface RecentItemQueryArgs {
  first: number;
}

type AnnotatedAlbum = Album & { kind: 'Album' };

type AnnotatedArtist = Artist & { kind: 'Artist' };

type AnnotatedRecentItem = AnnotatedAlbum | AnnotatedArtist;

type RecentItem = Album | Artist;

const annotateAlbum = (album: Album): AnnotatedAlbum => ({
  kind: 'Album',
  ...album,
});

const annotateArtist = (artist: Artist): AnnotatedArtist => ({
  kind: 'Artist',
  ...artist,
});

const getAnnotatedAlbums = () => [...albums.values()].map(annotateAlbum);

const getAnnotatedArtists = () => [...artists.values()].map(annotateArtist);

const getAnnotatedItems = (): AnnotatedRecentItem[] => [
  ...getAnnotatedAlbums(),
  ...getAnnotatedArtists(),
];

const getMostRecent = (
  allItems: AnnotatedRecentItem[],
  top: number,
  timeGetter: (item: AnnotatedRecentItem) => number
) => {
  allItems.sort((a: AnnotatedRecentItem, b: AnnotatedRecentItem) =>
    reverseCompare(timeGetter(a), timeGetter(b))
  );

  return allItems.slice(0, top);
};

export const recentlyAdded = (
  _: void,
  { first }: RecentItemQueryArgs
): RecentItem[] =>
  getMostRecent(getAnnotatedItems(), first, item => item.timeAdded);

export const recentlyPlayed = (
  _: void,
  { first }: RecentItemQueryArgs
): RecentItem[] =>
  getMostRecent(
    getAnnotatedItems().filter(item => !!item.lastPlayed),
    first,
    item => item.lastPlayed as number
  );

export const RecentItem = {
  __resolveType(item: AnnotatedRecentItem) {
    return item.kind;
  },
};
