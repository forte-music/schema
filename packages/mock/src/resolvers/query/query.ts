import { albums, artists, songs } from '../../models/index';
import { mustGet } from '../../utils';
import { itemsResolver } from './sort';
import { IResolverObject } from 'graphql-tools/dist/Interfaces';
import { RecentItem, recentlyAdded, recentlyPlayed } from './recentItems';

const itemResolver = <T>(map: Map<string, T>) => (
  _: void,
  { id }: { id: string }
): T => mustGet(map, id);

const queryResolvers = {
  Query: {
    album: itemResolver(albums),
    albums: itemsResolver(albums),

    artist: itemResolver(artists),
    artists: itemsResolver(artists),

    song: itemResolver(songs),
    songs: itemsResolver(songs),

    recentlyAdded,
    recentlyPlayed,
  },

  RecentItem,
} as IResolverObject;

export default queryResolvers;
