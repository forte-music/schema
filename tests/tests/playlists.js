import client from '../client';
import query from './playlist.graphql';

it('should resolve a query for data stemming from a playlist', async () => {
  const resp = await client.request(query);
  expect(resp).toMatchSnapshot();
});
