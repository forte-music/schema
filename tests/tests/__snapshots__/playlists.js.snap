// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should resolve a query for data stemming from a playlist 1`] = `
Object {
  "playlist": Object {
    "duration": 481,
    "id": "playlist:1",
    "items": Object {
      "edges": Array [
        Object {
          "node": Object {
            "id": "playlist:1:0",
            "song": Object {
              "album": Object {
                "artworkUrl": "http://is4.mzstatic.com/image/thumb/Music5/v4/08/da/96/08da9619-3f9b-7c95-60d1-6c18cfdd4dbd/source/600x600bb.jpg",
                "duration": 211,
                "id": "album:2",
                "name": "I'm That (Remix)",
                "releaseYear": 2015,
              },
              "artists": Array [
                Object {
                  "id": "artist:2",
                  "name": "R. City",
                },
                Object {
                  "id": "artist:22",
                  "name": "Azealia Banks",
                },
                Object {
                  "id": "artist:21",
                  "name": "Beenie Man",
                },
              ],
              "diskNumber": 1,
              "id": "song:2:1",
              "name": "I'm That (Remix)",
              "stats": Object {
                "id": "stats:2:1",
                "lastPlayed": 0,
                "liked": false,
                "playCount": 0,
              },
              "streamUrl": "/music/01 - I'm That... (Remix) [feat. Beenie Man & Azealia Banks].m4a",
              "trackNumber": 1,
            },
          },
        },
        Object {
          "node": Object {
            "id": "playlist:1:1",
            "song": Object {
              "album": Object {
                "artworkUrl": "https://upload.wikimedia.org/wikipedia/en/3/3f/AceHood_Bugatti.jpg",
                "duration": 270,
                "id": "album:3",
                "name": "Bugatti",
                "releaseYear": 2013,
              },
              "artists": Array [
                Object {
                  "id": "artist:24",
                  "name": "Future",
                },
                Object {
                  "id": "artist:3",
                  "name": "Ace Hood",
                },
                Object {
                  "id": "artist:23",
                  "name": "Rick Ross",
                },
              ],
              "diskNumber": 1,
              "id": "song:3:1",
              "name": "Bugatti",
              "stats": Object {
                "id": "stats:3:1",
                "lastPlayed": 0,
                "liked": false,
                "playCount": 0,
              },
              "streamUrl": "/music/01 - Bugatti (feat. Future & Rick Ross).flac",
              "trackNumber": 1,
            },
          },
        },
      ],
    },
    "name": "Quick Jam Session",
  },
}
`;
